heron/metricscachemgr/src/java/com/twitter/heron/metricscachemgr/metricscache/CacheCore.java

```diff
@@ -0,0 +1,571 @@
+    LOG.info("getRawMetrics " + startTime + " " + endTime);
+    Long startKey = cacheMetric.floorKey(startTime);
+    for (Long key = startKey != null ? startKey : cacheMetric.firstKey();
+         key != null && key <= endTime;
+         key = cacheMetric.higherKey(key)) {
+      LinkedList<MetricDatapoint> bucket = cacheMetric.get(key).get(bucketId);
+
+      if (bucket != null) {
+        for (MetricDatapoint datapoint : bucket) {
+          if (startTime <= datapoint.timestamp && datapoint.timestamp <= endTime) {
+            // per data point
+            metricValue.add(new MetricsCacheQueryUtils.MetricTimeRangeValue(
+                datapoint.timestamp, datapoint.timestamp, datapoint.value));
+          }
+        } // end bucket
+      }
+
+    } // end tree
+  }
+
+  // we assume the metric value is Double: compatible with tmaster
+  @SuppressWarnings("fallthrough")
+  private void getMinuteMetrics(List<MetricsCacheQueryUtils.MetricTimeRangeValue> metricValue,
+                                long startTime, long endTime, long bucketId,
+                                MetricsFilter.MetricAggregationType type) {
+    LOG.info("getMinuteMetrics " + startTime + " " + endTime);
+    Long startKey = cacheMetric.floorKey(startTime);
+    for (Long key = startKey != null ? startKey : cacheMetric.firstKey();
+         key != null && key <= endTime;
+         key = cacheMetric.higherKey(key)) {
+      LinkedList<MetricDatapoint> bucket = cacheMetric.get(key).get(bucketId);
+
+      if (bucket != null) {
+        // per bucket
+        long countAvg = 0;
+
+        // prepare range value
+        long outStartTime = Long.MAX_VALUE;
+        long outEndTime = 0;
+        String outValue = null;
+
+        double result = 0;
+        for (MetricDatapoint datapoint : bucket) {
+          if (startTime <= datapoint.timestamp && datapoint.timestamp <= endTime) {
+            switch (type) {
+              case AVG:
+                countAvg++;
+              case SUM:
+                result += Double.parseDouble(datapoint.value);
+                break;
+              case LAST:
+                if (outEndTime < datapoint.timestamp) {
+                  outValue = datapoint.value;
+                }
+                break;
+              case UNKNOWN:
+              default:
+                LOG.warning(
+                    "Unknown metric type, CacheCore does not know how to aggregate " + type);
+                return;
+            }
+            outStartTime = Math.min(outStartTime, datapoint.timestamp);
+            outEndTime = Math.max(outEndTime, datapoint.timestamp);
+          }
+        } // end bucket
+
+        if (type.equals(MetricsFilter.MetricAggregationType.AVG)) {
+          outValue = String.valueOf(result / countAvg);
+        } else if (type.equals(MetricsFilter.MetricAggregationType.SUM)) {
+          outValue = String.valueOf(result);
+        }
+        if (outValue != null) {
```

Comment:
Given LAST type, if the metric timestamp is out of query range, the outValue is null.
It indicates the metric is not found for the given query in the bucket, which is not an error condition.